Add Flex example page

tzujenchanmbd opened the pull request:
The current Flex spec proposal contains some aspects that are difficult to fully explain in text. Also, we prefer not to overload the spec with numerous data examples (tables), which could make it difficult to read. Therefore, we have added this data examples page. Some descriptions in the spec will directly link to the corresponding examples on this page.

The Flex specification proposal is about to enter the voting stage. If the vote passes successfully, we plan to release this page after the vote as a reference information for the spec.

This PR represents the initial release of the content, and we plan to continue improving the content and layout in the future. Please review this PR and feel free to share any questions/suggestions/thoughts here.

tzujenchanmbd committed:
Create flex.md

Outline / location_group example / routing behavior example

tzujenchanmbd committed:
editorial changes for consistency

tzujenchanmbd committed:
Add within a single zone example

tzujenchanmbd committed:
Editorial changes

tzujenchanmbd committed:
Add between 2 zones example

tzujenchanmbd committed:
add deviated route example

tzujenchanmbd committed:
editorial update

tzujenchanmbd committed:
Add files via upload

tzujenchanmbd committed:
Add files via upload

tzujenchanmbd committed:
add .zip and .png resources

westontrillium commented on docs/schedule/examples/flex.md:
```diff
@@ -0,0 +1,287 @@
+### Define stop_times
+- For fixed stops, define fields such as `arrival_time`, `departure_time`, and `stop_id` in a manner similar to normal bus routes.
+- Between fixed stops, define the zones where deviation is allowed.
+- `pickup_type = 1` and `drop_off_type = 3` indicates that pre-booking is required for boarding in the deviation zone, and riders must coordinate with the driver for alighting in the zone.
```
For this example service, deviated pickup is actually not allowed (`pickup_type = 1`). Riders who are already on board the vehicle may coordinate with the driver for a deviated drop-off *only* (`drop_off_type = 3`).

Sergiodero replied:
Thanks @westontrillium! We've updated the text to reflect this and provided additional clarification on the service characteristics for this example.

Sergiodero committed:
Fixed tables, updated text & added in nav

- Fixed table formatting.
- Modified text to reflect proper description of pickup_type=1 in Deviated route example.
- Added section to navigation in mkdocs.yml

Sergiodero reviewed (approved):
I've reviewed the content and made some adjustments, overall it looks good to me! changes made in the last commit are as follow:

- Added section to the navigation menu in mkdocs.yml,
- Fixed table and bullet point formatting,
- Made corrections to text for Deviated route section based on @westontrillium feedback.

@isabelle-dr do you mind taking one final look at the changes?

Thanks!

isabelle-dr commented on mkdocs.yml:
```diff
@@ -558,6 +558,7 @@ nav:
       - Translations: schedule/examples/translations.md
       - Feed information: schedule/examples/feed-info.md
       - Dataset attributions: schedule/examples/attributions.md
+      - Flex: schedule/examples/flex.md
```
I think I'd rename this to Demand responsive services, I think it's more widely understood.

Sergiodero replied:
Updated title in page and in navigation panel

isabelle-dr replied:
Nit: Would it be worth it to also rename `flex.md` itself to `demand-responsive.md`?

isabelle-dr commented on docs/schedule/examples/flex.md:
```diff
@@ -0,0 +1,326 @@
+
+<hr>
+
+Flex is a GTFS extension project that aims to facilitate discoverability of Demand Responsive Transportation Services.
```
Maybe add "which was adopted officially into the spec in X 2024" (which we can update when the vote has passed)?

Sergiodero replied:
Updated

isabelle-dr commented on docs/schedule/examples/flex.md:
```diff
@@ -0,0 +1,326 @@
+flächenrufbus_angermünde_weekdays | 1 | 60 | Anmeldung mind. 60min vorher erforderlich, per Anruf zwischen 08:00 und 24:00 möglich, oder online rund um die Uhr | +49 3332 442 755 | https://uvg-online.com/rufbus-angermuende/ | https://uvg.tdimo.net/bapp/#/astBuchungenView
+flächenrufbus_angermünde_weekends | 1 | 60 | 1€ Komfortzuschlag pro Person; Anmeldung mind. 60min vorher erforderlich, per Anruf zwischen 08:00 und 24:00 möglich, oder online rund um die Uhr | +49 3332 442 755 | https://uvg-online.com/rufbus-angermuende/ | https://uvg.tdimo.net/bapp/#/astBuchungenView
+
+### Define stop_times
```
I think we normally use "stop times" or "stop-times.txt". I would replace with "Define stop times".

Sergiodero replied:
Harmonized text to use "stop times"

isabelle-dr commented on docs/schedule/examples/flex.md:
```diff
@@ -0,0 +1,326 @@
+-- | -- | -- | -- | -- | -- | -- | -- | --
+booking_route_74362 | 2 | 14 | 8:00:00 | 1 | 15:00:00 | Brown County Heartland Express provides door-to-door on-demand transportation. To request a ride, call 1-507-359-2717 or 1-800-707-2717 by 3pm at least one business day ahead of your trip. | (507) 359-2717 | https://www.co.brown.mn.us/heartland-express-transit
+
+### Define stop_times
```
Similar comment than [here](https://github.com/MobilityData/gtfs.org/pull/301/files#r1489945288)

isabelle-dr commented on docs/schedule/examples/flex.md:
```diff
@@ -0,0 +1,326 @@
+
+`service_id = c_67295_b_77497_d_31` refers to weekdays, `service_id = c_67295_b_77497_d_64` refers to Sunday.
+
+### Define zones (GeoJSON locations)
```
We're in the "On-demand services within a single zone" section, maybe this should be "Define zone" instead? (in the singular form)

Sergiodero replied:
Updated

isabelle-dr commented on docs/schedule/examples/flex.md:
```diff
@@ -0,0 +1,326 @@
+
+`area_715` refers to New Ulm City zone, `area_708` refers to Bronw County zone. 
+
+## On-demand services between multiple zones
```
Maybe "On-demand services across multiple zones"?

Sergiodero replied:
Updated text

isabelle-dr commented on docs/schedule/examples/flex.md:
```diff
@@ -0,0 +1,326 @@
+
+### Define stop_times
+
+- The following data indicates that boarding is only allowed in one zone and alighting is only allowed in another zone. Boarding and alighting in the same zone are not allowed.
```
I think we can remove the bullet point for consistency with other sections.

Sergiodero replied:
Updated format

isabelle-dr commented on docs/schedule/examples/flex.md:
```diff
@@ -0,0 +1,326 @@
+
+### Define stop_times
+
+- The operating hours are defined using the `start_pickup_drop_off_window` and `end_pickup_drop_off_window` fields.
```
I think we can remove the first level of bullet point for consistency with other sections.

Sergiodero replied:
Updated format

isabelle-dr commented on docs/schedule/examples/flex.md:
```diff
@@ -0,0 +1,326 @@
+
+### Define stop_times
+
+- For fixed stops, define fields such as `arrival_time`, `departure_time`, and `stop_id` in a manner similar to normal bus routes.
```
I think we can remove the bullet points in this section.

Sergiodero replied:
Updated format

isabelle-dr commented on docs/schedule/examples/flex.md:
```diff
@@ -0,0 +1,326 @@
+
+### Define zones (GeoJSON location)
+
+Using locations.geojson to define zones for deviated route. Typically, deviations are limited to keep the service on schedule. Therefore, as the vehicle travels, the deviation area between each fixed stop may vary accordingly. The area for route deviation may look like the image below:
```
When merged, add the link to the file

Sergiodero replied:
Added link

isabelle-dr replied:
It doesn't seem to render properly on markdown 🤔.
<img width="457" alt="Screenshot 2024-04-05 at 1 39 47 PM" src="https://github.com/MobilityData/gtfs.org/assets/63653518/90adb9d5-5635-4ec1-ada7-01b761f9f1f8">

We're still missing links in a few other places where locations.geojson is referenced, and one place where booking_rules.txt is referenced.

isabelle-dr commented on docs/schedule/examples/flex.md:
```diff
@@ -0,0 +1,326 @@
+
+## On-demand services between multiple zones
+
+Some demand-responsive services operate between multiple distinct zones, where riders can book pickups at any point A within one area and drop-offs at any point B within another area. For example, [Minnesota River Valley Transit](https://www.saintpetermn.gov/330/Dial-a-Ride) offers on-demand services between Saint Peter and Kasota cities:
```
A suggestion to make this paragraph clearer:

Some demand-responsive services operate **across** multiple zones, where riders can book pickups at any **location** within one area and drop-offs at any **location** within another area.

Also, here we use pickup/drop-off, and later in the page, we talk about using board/alight. 
I understand pickup/dropoff references the driver's perspective and board/alight references the rider's perspective, but if there is a way to only use one, it would increase clarity (e. g. riders must be picked up and dropped off at specific locations)

Sergiodero replied:
Updated text, rest of document has been harmonized to use pickup/dropoff

isabelle-dr commented on docs/schedule/examples/flex.md:
```diff
@@ -0,0 +1,326 @@
+
+- Weekdays:
+  - 8:00 AM - 5:00 PM
+  - 6:15 AM – 5:45 PM (New Ulm Only)
```
It's not clear at first what New Ulm is. 
I would change to "New Ulm zone only" in this line and below.

Sergiodero replied:
Updated text

isabelle-dr commented on docs/schedule/examples/flex.md:
```diff
@@ -0,0 +1,326 @@
+  - 6:15 AM – 5:45 PM (New Ulm Only)
+- Sunday: 8:00 AM - Noon (New Ulm Only)
+
+The New Ulm city zone is included within the Brown County zone. To avoid the "zone overlap constraint" issue, Heartland Express can be defined with four trips:
```
I think contained would be the appropriate term.

The New Ulm city zone is **contained** within the Brown County zone.

Also, it's not clear to me what the "zone overlap constraint issue is" from reading this.

Sergiodero replied:
Updated text, added link to provide more information on the zone overlap constraint issue

isabelle-dr commented on docs/schedule/examples/flex.md:
```diff
@@ -0,0 +1,326 @@
+
+### Define trips
+
+Route 476 offers on-demand services between each stop in the Angermünde region, including two trips(services): one for weekdays and the other for weekends. 
```
I am a little confused by 
- the "including": do they offer this only for two trips that run during the night, or do they just operate two trips in total?
- the "trips(services)", given that trips and services are different concepts in gtfs.

Suggestion (assuming what the above means here, feel free to update):
_Route 476 offers on-demand services between each stop in the Angermünde region. They operate two services that each have one single trip associated._

Sergiodero replied:
Updated text with slight changes

isabelle-dr commented on docs/schedule/examples/flex.md:
```diff
@@ -0,0 +1,326 @@
+
+### Define booking rules
+
+- The 476 route service requires booking at least one hour in advance. Using `booking_type = 1` indicates that the service requires up to same-day booking with advance notice. The `prior_notice_duration_min = 60` indicates a requirement for booking at least 60 minutes in advance.
```
Again I'd remove the bullet points for consistency, this is not really a list.

Sergiodero replied:
Updated format

isabelle-dr commented on docs/schedule/examples/flex.md:
```diff
@@ -0,0 +1,326 @@
+
+### Define stop_times
+
+- The 476 route operates from 5:30 PM to 10:00 PM on weekdays and from 8:00 AM to 10:00 PM on weekends. The operating hours are defined using the `start_pickup_drop_off_window` and `end_pickup_drop_off_window` fields.
```
I'd remove the first level of bullet points

Sergiodero replied:
Updated format

isabelle-dr commented on docs/schedule/examples/flex.md:
```diff
@@ -0,0 +1,326 @@
+
+## Deviated route
+
+"Route deviation" refers to services where the vehicle follows a fixed route with a set sequence of stops but has the flexibility to deviate from this route to pick up or drop off riders between stops. Typically, deviations are limited to keep the service on schedule, and advance booking is required for deviated pickups. 
```
Suggestion to increase clarity:

Typically, deviations are limited to **maintain service punctuality**, and **prior** booking is required for deviated pickups **and dropoffs**.

Sergiodero replied:
Updated text

isabelle-dr commented on docs/schedule/examples/flex.md:
```diff
@@ -0,0 +1,326 @@
+
+### Define trips
+
+Since this type of service still involves a series of fixed stops and a fixed schedule, defining trips is similar to normal fixed-route bus services. It requires defining trips for each service throughout the day.
```
Suggestion to increase clarify:

It requires defining the **trips served by each route throughout all relevant service periods.**

Sergiodero replied:
Updated text

isabelle-dr commented on docs/schedule/examples/flex.md:
```diff
@@ -0,0 +1,326 @@
+
+## Routing behavior
+
+### Ignoring intermediate stop_times records with pickup/drop-off Windows
```
use stop_times.txt or stop times

Sergiodero replied:
Changed to "stop times"

isabelle-dr reviewed (commented):
Thank you for working on this @tzujenchanmbd, @westontrillium and @Sergiodero!

I left a variety of minor comments, overall this looks very good. 
I suggest we merge this in gtfs.org when the PR is merged on google/transit.

Sergiodero committed:
Corrections on flex examples

Sergiodero commented:
Hi there! @isabelle-dr's comments have been addresses with the latest commit, @tzujenchanmbd could you do a final review?

isabelle-dr commented on docs/schedule/examples/flex.md:
```diff
@@ -0,0 +1,324 @@
+
+<hr>
+
+Flex is a GTFS extension project which was adopted officially into the GTFS specification in March 2024, its aims to facilitate discoverability of Demand Responsive Transportation Services.
```
Nit: suggestion to link "demand responsive" and "GTFS Flex" together a little more clearly. 

_Demand-responsive services can be modeled in GTFS using the GTFS Flex spec extension, which was adopted in March 2024._

isabelle-dr commented on docs/schedule/examples/flex.md:
```diff
@@ -0,0 +1,324 @@
+
+## On-demand services across multiple zones
+
+Some demand-responsive services operate across multiple distinct zones, where riders can book pickups at any location A within one area and drop-offs at any location within another area. For example, [Minnesota River Valley Transit](https://www.saintpetermn.gov/330/Dial-a-Ride) offers on-demand services between Saint Peter and Kasota cities:
```
Nit: do we need the "A"?

isabelle-dr reviewed (approved):
Thanks for the updates. Minor final comments & we're good to merge 🚀. Thank you to everyone who participated in drafting this page!

isabelle-dr commented:
This might be for another PR, but https://gtfs.org/extensions/flex/ is one of the top 5 pages visited on gtfs.org. 
Now that it has been adopted, it would be nice to either:
- add a mention that it has been adopted, reference the new data example page
- have a plan to archive it, maybe by reallocating its content to other places on the site (GTFS features and/or add more this example page?)

isabelle-dr commented on docs/schedule/examples/flex.md:
```diff
@@ -0,0 +1,324 @@
+  - 6:15 AM – 5:45 PM (New Ulm zone only)
+- Sunday: 8:00 AM - Noon (New Ulm zone only)
+
+The New Ulm city zone is contained within the Brown County zone. To avoid the ["zone overlap constraint"](/schedule/examples/flex/#zone-overlap-constraint) issue, Heartland Express can be defined with four trips:
```
This gives a 404; I think the appropriate way to refer to something in the same page is to use `(#zone-overlap-constraint)`

Sergiodero replied:
Updated! the previous link was working for me when I displayed the changes locally but change it to the suggested link, it should be ok now!

Sergiodero committed:
Update flex.md

Fix zone overlap constrain link
